### ISSUE TITLE
Does it work under Windows/Cygwin?

### ISSUE THREAD

**Issue by tkossak** [xonsh/xonsh#514]
Windows 7 x64
Cygwin x86
Python3 3.4.3

I installed xonsh with ```pip3 install xonsh``` and when I try to run it with ```xonsh``` I get the error:
```
Exception in thread Thread-2:
Traceback (most recent call last):
  File "/usr/lib/python3.4/threading.py", line 920, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.4/site-packages/xonsh/readline_shell.py", line 281, in run
    RL_LIB.history_set_pos(i)
  File "/usr/lib/python3.4/ctypes/__init__.py", line 364, in __getattr__
    func = self.__getitem__(name)
  File "/usr/lib/python3.4/ctypes/__init__.py", line 369, in __getitem__
    func = self._FuncPtr((name_or_ordinal, self))
AttributeError: function 'history_set_pos' not found (ďż˝[ďż˝ďż˝ďż˝%ďż˝ďż˝Pďż˝ďż˝ďż˝)
```

**Comment by scopatz** [comment]
You may need to install gnureadline or use prompt-toolkit instead.

**Comment by scopatz** [comment]
Thanks for reporting @tkossak

**Comment by tkossak** [comment]
```pip3 install gnureadline``` gave me errors (gcc can't find lncurses) but I
managed to run xonsh with ```prompt-toolkit```. The shell starts with no errors, but
when I run any shell command (ie: ls, vim, echo, etc.) I get the error:

```
Traceback (most recent call last):
  File "/usr/lib/python3.4/site-packages/xonsh/base_shell.py", line 126, in default
    self.execer.exec(code, mode='single', glbs=self.ctx)  # no locals
  File "/usr/lib/python3.4/site-packages/xonsh/execer.py", line 126, in exec
    return exec(code, glbs, locs)
  File "<xonsh-code>", line 1, in <module>
  File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 636, in subproc_uncaptured
    return run_subproc(cmds, captured=False)
  File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 606, in run_subproc
    wait_for_active_job()
  File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 110, in wait_for_active_job
    _give_terminal_to(pgrp)
  File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 88, in _give_terminal_to
    oldmask = signal.pthread_sigmask(signal.SIG_BLOCK, _block_when_giving)
AttributeError: 'module' object has no attribute 'pthread_sigmask'
```

**Comment by scopatz** [comment]
What terminal emulator are you running in? Also are you using cygwin's Python?  Also, what is the output of:

```console
import xonsh.tools
print(xonsh.tools.ON_WINDOWS, xonsh.tools.ON_LINUX, xonsh.tools.ON_POSIX)
```

I bet this is xonsh becoming confused about whether you are on windows or posix.

**Comment by scopatz** [comment]
This might be the first report of anyone running in cygwin.

**Comment by tkossak** [comment]
hehe, i'm forced to use windows at work so the least I can do to make it behave like real os is installing cygwin :)

My terminal emulator: mintty + tmux (and everything inside Cmder/ConEmu)
I'm using cygwin's python/pip, not window's ones. On bash:
```
$ which pip pip2 pip3 python python2 python3
/usr/bin/pip
/usr/bin/pip2
/usr/bin/pip3
/usr/bin/python
/usr/bin/python2
/usr/bin/python3

$ pip --version
pip 7.1.2 from /usr/lib/python2.7/site-packages (python 2.7)
$ pip2 --version
pip 7.1.2 from /usr/lib/python2.7/site-packages (python 2.7)
$ pip3 --version
pip 7.1.2 from /usr/lib/python3.4/site-packages (python 3.4)
$ python --version
Python 2.7.10
$ python2 --version
Python 2.7.10
$ python3 --version
Python 3.4.3
```

**on xonsh:**

```
$ which pip pip2 pip3 python python2 python3
Traceback (most recent call last):
File "/usr/lib/python3.4/site-packages/xonsh/base_shell.py", line 126, in default
self.execer.exec(code, mode='single', glbs=self.ctx) # no locals
File "/usr/lib/python3.4/site-packages/xonsh/execer.py", line 126, in exec
return exec(code, glbs, locs)
File "<xonsh-code>", line 1, in <module>
File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 636, in subproc_uncaptured
return run_subproc(cmds, captured=False)
File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 606, in run_subproc
wait_for_active_job()
File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 110, in wait_for_active_job
_give_terminal_to(pgrp)
File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 88, in _give_terminal_to
oldmask = signal.pthread_sigmask(signal.SIG_BLOCK, _block_when_giving)
AttributeError: 'module' object has no attribute 'pthread_sigmask'
/usr/bin/pip
/usr/bin/pip2
/usr/bin/pip3
/usr/bin/python
/usr/bin/python2
/usr/bin/python3


$ pip --version
Traceback (most recent call last):
File "/usr/lib/python3.4/site-packages/xonsh/base_shell.py", line 126, in default
self.execer.exec(code, mode='single', glbs=self.ctx) # no locals
File "/usr/lib/python3.4/site-packages/xonsh/execer.py", line 126, in exec
return exec(code, glbs, locs)
File "<xonsh-code>", line 1, in <module>
File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 636, in subproc_uncaptured
return run_subproc(cmds, captured=False)
File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 606, in run_subproc
wait_for_active_job()
File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 110, in wait_for_active_job
_give_terminal_to(pgrp)
File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 88, in _give_terminal_to
oldmask = signal.pthread_sigmask(signal.SIG_BLOCK, _block_when_giving)
AttributeError: 'module' object has no attribute 'pthread_sigmask'
pip 7.1.2 from /usr/lib/python2.7/site-packages (python 2.7)

$ pip2 --version
Traceback (most recent call last):
File "/usr/lib/python3.4/site-packages/xonsh/base_shell.py", line 126, in default
self.execer.exec(code, mode='single', glbs=self.ctx) # no locals
File "/usr/lib/python3.4/site-packages/xonsh/execer.py", line 126, in exec
return exec(code, glbs, locs)
File "<xonsh-code>", line 1, in <module>
File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 636, in subproc_uncaptured
return run_subproc(cmds, captured=False)
File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 606, in run_subproc
wait_for_active_job()
File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 110, in wait_for_active_job
_give_terminal_to(pgrp)
File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 88, in _give_terminal_to
oldmask = signal.pthread_sigmask(signal.SIG_BLOCK, _block_when_giving)
AttributeError: 'module' object has no attribute 'pthread_sigmask'
pip 7.1.2 from /usr/lib/python2.7/site-packages (python 2.7)

$ pip3 --version
Traceback (most recent call last):
File "/usr/lib/python3.4/site-packages/xonsh/base_shell.py", line 126, in default
self.execer.exec(code, mode='single', glbs=self.ctx) # no locals
File "/usr/lib/python3.4/site-packages/xonsh/execer.py", line 126, in exec
return exec(code, glbs, locs)
File "<xonsh-code>", line 1, in <module>
File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 636, in subproc_uncaptured
return run_subproc(cmds, captured=False)
File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 606, in run_subproc
wait_for_active_job()
File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 110, in wait_for_active_job
_give_terminal_to(pgrp)
File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 88, in _give_terminal_to
oldmask = signal.pthread_sigmask(signal.SIG_BLOCK, _block_when_giving)
AttributeError: 'module' object has no attribute 'pthread_sigmask'
pip 7.1.2 from /usr/lib/python3.4/site-packages (python 3.4)

$ python --version
Traceback (most recent call last):
File "/usr/lib/python3.4/site-packages/xonsh/base_shell.py", line 126, in default
self.execer.exec(code, mode='single', glbs=self.ctx) # no locals
Python 2.7.10
File "/usr/lib/python3.4/site-packages/xonsh/execer.py", line 126, in exec
return exec(code, glbs, locs)
File "<xonsh-code>", line 1, in <module>
File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 636, in subproc_uncaptured
return run_subproc(cmds, captured=False)
File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 606, in run_subproc
wait_for_active_job()
File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 110, in wait_for_active_job
_give_terminal_to(pgrp)
File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 88, in _give_terminal_to
oldmask = signal.pthread_sigmask(signal.SIG_BLOCK, _block_when_giving)
AttributeError: 'module' object has no attribute 'pthread_sigmask'

$ python2 --version
Traceback (most recent call last):
Python 2.7.10
File "/usr/lib/python3.4/site-packages/xonsh/base_shell.py", line 126, in default
self.execer.exec(code, mode='single', glbs=self.ctx) # no locals
File "/usr/lib/python3.4/site-packages/xonsh/execer.py", line 126, in exec
return exec(code, glbs, locs)
File "<xonsh-code>", line 1, in <module>
File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 636, in subproc_uncaptured
return run_subproc(cmds, captured=False)
File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 606, in run_subproc
wait_for_active_job()
File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 110, in wait_for_active_job
_give_terminal_to(pgrp)
File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 88, in _give_terminal_to
oldmask = signal.pthread_sigmask(signal.SIG_BLOCK, _block_when_giving)
AttributeError: 'module' object has no attribute 'pthread_sigmask'

$ python3 --version
Traceback (most recent call last):
File "/usr/lib/python3.4/site-packages/xonsh/base_shell.py", line 126, in default
self.execer.exec(code, mode='single', glbs=self.ctx) # no locals
File "/usr/lib/python3.4/site-packages/xonsh/execer.py", line 126, in exec
return exec(code, glbs, locs)
File "<xonsh-code>", line 1, in <module>
File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 636, in subproc_uncaptured
return run_subproc(cmds, captured=False)
File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 606, in run_subproc
wait_for_active_job()
File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 110, in wait_for_active_job
_give_terminal_to(pgrp)
File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 88, in _give_terminal_to
oldmask = signal.pthread_sigmask(signal.SIG_BLOCK, _block_when_giving)
AttributeError: 'module' object has no attribute 'pthread_sigmask'
Python 3.4.3
```

**and the code you asked:**
```
$ import xonsh.tools
$ print(xonsh.tools.ON_WINDOWS, xonsh.tools.ON_LINUX, xonsh.tools.ON_POSIX)
Traceback (most recent call last):
File "/usr/lib/python3.4/site-packages/xonsh/base_shell.py", line 158, in push
locs=self.ctx)
File "/usr/lib/python3.4/site-packages/xonsh/execer.py", line 94, in compile
tree = self.parse(input, ctx, mode=mode)
File "/usr/lib/python3.4/site-packages/xonsh/execer.py", line 69, in parse
tree = self._parse_ctx_free(input, mode=mode)
File "/usr/lib/python3.4/site-packages/xonsh/execer.py", line 150, in _parse_ctx_free
debug_level=self.debug_level)
File "/usr/lib/python3.4/site-packages/xonsh/parser.py", line 263, in parse
tree = self.parser.parse(input=s, lexer=self.lexer, debug=debug_level)
File "/usr/lib/python3.4/site-packages/ply/yacc.py", line 331, in parse
return self.parseopt_notrack(input, lexer, debug, tracking, tokenfunc)
File "/usr/lib/python3.4/site-packages/ply/yacc.py", line 1106, in parseopt_notrack
p.callable(pslice)
File "/usr/lib/python3.4/site-packages/xonsh/parser.py", line 2098, in p_arglist
for arg in p1:
TypeError: 'Attribute' object is not iterable
```

**Comment by scopatz** [comment]
What version of ply are you on? Are you sure you aren't running xonsh in python 2?

**Comment by tkossak** [comment]
ran inside xonsh:
```
$ import sys
$ print(sys.version)
3.4.3 (default, May  5 2015, 17:04:32)
[GCC 4.9.2]
$ import ply
$ ply.__version__
'3.7'
```

**Comment by scopatz** [comment]
OK. I have no idea what is going on here.  Can you try asking on the mailing list or maybe stack overflow.  Unfortunately, I don't have access to a windows system to test things out myself.  But maybe other xonsh-windows users could try it out on cygwin.

**Comment by tildebyte** [comment]
I don't get any errors; I just don't get anything. I run `xonsh`, and I get nothing - no prompt, no exit back to bash.

CTRL-C to exit gives me a [traceback](https://gist.github.com/tildebyte/0d59e70bd55e52c97f10#file-traceback-py).

RE: "gnureadline or use prompt-toolkit instead" - I installed `prompt-toolkit`, but I have no idea what xonsh is using.

Latest (as of today) cygwin x64.
* "Pure" Cygwin, i.e. gnome-terminal 3.18.2 on Cyg-X.
* Windows 7 Enterprise sp1.
* Python 3.4.3 virtual environment.


````bash
(xonsh) foo@foo.com ~/
$ python
Python 3.4.3 (default, May  5 2015, 17:58:45) 
[GCC 4.9.2] on cygwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import xonsh.tools
>>> print(xonsh.tools.ON_WINDOWS, xonsh.tools.ON_LINUX, xonsh.tools.ON_POSIX)
False False True
>>> import ply
>>> ply.__version__
'3.7'
````
As far as xonsh being confused about the platform it's running on, [this](https://gist.github.com/tildebyte/0d59e70bd55e52c97f10#file-xonsh-py) is the xonsh command I ended up with. Interesting.

**Comment by scopatz** [comment]
Hi @tildebyte - thanks for reporting.  Can you run the `xonfig` command and report back what it tells you?  This is super weird.  If you need to run it manually, you should be able to do:

```python
from xonsh import xonfig
xonfig.main([])
```

**Comment by tildebyte** [comment]
````
Python 3.4.3 (default, May  5 2015, 17:58:45) 
[GCC 4.9.2] on cygwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from xonsh import xonfig
>>> xonfig.main([])
+------------------+-------+
| xonsh            | 0.2.6 |
| Python           | 3.4.3 |
| PLY              | 3.7   |
| have readline    | True  |
| prompt toolkit   | 0.57  |
| pygments         | 2.1.1 |
| on posix         | True  |
| on linux         | False |
| on arch          | False |
| on windows       | False |
| on mac           | False |
| are root user    | False |
| default encoding | utf-8 |
+------------------+-------+
````

**Comment by scopatz** [comment]
Hi @tildebyte - OK it looks like it is getting stuck in the foreign shell phase on startup.  Can you try adding a config file that does not load in any foreign shells, as described http://xon.sh/xonshconfig.html

Basically make `~/.config/xonsh/config.json` look like:

```json
{}
```

We should probably remove bash from the default foreign shells on posix.

**Comment by tildebyte** [comment]
Hmm. Neither `{}` nor (as from the docs) `{"foreign_shells": []}` help.

**Comment by scopatz** [comment]
Hmm. Is your `$BASH_COMPLETIONS` variable set to an empty list?

**Comment by scopatz** [comment]
or tuple.

**Comment by tildebyte** [comment]
`$BASH_COMPLETIONS` is unset.

**Comment by scopatz** [comment]
@tkossak could you try running xonsh through a debugger or profiler to see where the problem is coming from?

**Comment by scopatz** [comment]
I feel like the error is probably something trivial if we could identify what it is.

**Comment by tkossak** [comment]
When I checked xonsh today I have the same results as @tildebyte - after running `xonsh` i get no errors/no prompt/no exit back to bash. When i press CTRL+C i get:
```
Traceback (most recent call last):
  File "/usr/lib/python3.4/runpy.py", line 170, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.4/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/usr/lib/python3.4/trace.py", line 858, in <module>
    main()
  File "/usr/lib/python3.4/trace.py", line 804, in main
    t.runctx(code, globs, globs)
  File "/usr/lib/python3.4/trace.py", line 510, in runctx
    exec(cmd, globals, locals)
  File "/usr/bin/xonsh", line 11, in <module>
    sys.exit(main())
  File "/usr/lib/python3.4/site-packages/xonsh/main.py", line 166, in main
    args = premain(argv)
  File "/usr/lib/python3.4/site-packages/xonsh/main.py", line 153, in premain
    shell = builtins.__xonsh_shell__ = Shell(**shell_kwargs)
  File "/usr/lib/python3.4/site-packages/xonsh/shell.py", line 61, in __init__
    self._init_environ(ctx, config, rc)
  File "/usr/lib/python3.4/site-packages/xonsh/shell.py", line 99, in _init_environ
    self.execer = Execer(config=config)
  File "/usr/lib/python3.4/site-packages/xonsh/execer.py", line 40, in __init__
    load_builtins(execer=self, config=config)
  File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 675, in load_builtins
    builtins.__xonsh_env__ = ENV = Env(default_env(config=config))
  File "/usr/lib/python3.4/site-packages/xonsh/environ.py", line 1099, in default_env
    issue_warning=False))
  File "/usr/lib/python3.4/site-packages/xonsh/foreign_shells.py", line 419, in load_foreign_envs
    shenv, _ = foreign_shell_data(**shell)
  File "/usr/lib/python3.4/functools.py", line 472, in wrapper
    result = user_function(*args, **kwds)
  File "/usr/lib/python3.4/site-packages/xonsh/foreign_shells.py", line 173, in foreign_shell_data
    universal_newlines=True)
  File "/usr/lib/python3.4/subprocess.py", line 609, in check_output
    output, unused_err = process.communicate(inputdata, timeout=timeout)
  File "/usr/lib/python3.4/subprocess.py", line 960, in communicate
    stdout, stderr = self._communicate(input, endtime, timeout)
  File "/usr/lib/python3.4/subprocess.py", line 1617, in _communicate
    ready = selector.select(timeout)
  File "/usr/lib/python3.4/selectors.py", line 367, in select
    fd_event_list = self._poll.poll(timeout)
KeyboardInterrupt
Exception ignored in: <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>
BrokenPipeError: [Errno 32] Broken pipe
```

`xonfig.main([])` returns garbled string:
```
Python 3.4.3 (default, May  5 2015, 17:58:45)
[GCC 4.9.2] on cygwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from xonsh import xonfig
>>> xonfig.main([])
'+------------------+-------+\n| xonsh            | 0.2.6 |\n| Python           | 3.4.3 |\n| PLY              | 3.7   |\n| have readline    | True  |\n| prompt toolkit   | False |\n| pygments         | 2.0.2 |\n| on posix         | True  |\n| on linux         | False |\n| on arch          | False |\n| on windows       | False |\n| on mac           | False |\n| are root user    | False |\n| default encoding | utf-8 |\n+------------------+-------+\n'
>>>
```

Which after replacing `\n` with actual new lines looks like:
```

Python 3.4.3 (default, May  5 2015, 17:58:45)
[GCC 4.9.2] on cygwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from xonsh import xonfig
>>> xonfig.main([])
'+------------------+-------+
| xonsh            | 0.2.6 |
| Python           | 3.4.3 |
| PLY              | 3.7   |
| have readline    | True  |
| prompt toolkit   | False |
| pygments         | 2.0.2 |
| on posix         | True  |
| on linux         | False |
| on arch          | False |
| on windows       | False |
| on mac           | False |
| are root user    | False |
| default encoding | utf-8 |
+------------------+-------+
'
>>>
```

I tried running trace (using python `trace` module with --trace option) and I got this 12 MB file: https://gist.github.com/tkossak/3705a380664e33439385

Everything was run in Cygwin x64 (if it matters at all).

Hope it helps.

**Comment by scopatz** [comment]
Ok, not sure if this will do anything, but maybe you could try setting [`$FORCE_POSIX_PATHS = True`](http://xon.sh/envvars.html#force-posix-paths) if you haven't already.

**Comment by scopatz** [comment]
This seems to be getting hung up on running path manipulation while running subprocesses.  In particular, it is not happy with:

```
environ.py(125):     xcd = os.path.join(env.get('XDG_CONFIG_HOME'), 'xonsh')
```

**Comment by tkossak** [comment]
Running `export FORCE_POSIX_PATHS=True` before running `xonsh` didn't change anything.
I don't have XDG_CONFIG_HOME environmental variable but i tried `export XDG_CONFIG_HOME=/home/<USER>/.config` with no luck either.

**Comment by OllieTerrance** [comment]
I'm in a similar position here, though I don't seem to be getting very far into the startup.  Running `xonsh` shows no output.  Running via verbose Python:

```
$ python -v
...
Python 3.4.3 (default, May  5 2015, 17:58:45)
[GCC 4.9.2] on cygwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from xonsh import main
...
>>> main.main()
# /usr/lib/python3.4/site-packages/xonsh/__pycache__/parser_table.cpython-34.pyc matches /usr/lib/python3.4/site-packages/xonsh/parser_table.py
# code object from '/usr/lib/python3.4/site-packages/xonsh/__pycache__/parser_table.cpython-34.pyc'
import 'xonsh.parser_table' # <_frozen_importlib.SourceFileLoader object at 0x6ffff129b38>
```

At that point it just hangs indefinitely.  I have tried the suggestions above, but I suspect this isn't getting far enough for those to be the problem.

However, piping to stdin seems to bring me up to the error noted in the previous comments (note commands are being run, the output ending up amongst the stack trace):

```
$ xonsh <<EOF
ls
EOF
/usr/lib/python3.4/site-packages/xonsh/shell.py:97: RuntimeWarning: prompt-toolkit version < v0.57 and may not work as expected. Please update.
  warn(msg, RuntimeWarning)
Traceback (most recent call last):
  File "/usr/bin/xonsh", line 11, in <module>
    sys.exit(main())
  File "/usr/lib/python3.4/site-packages/xonsh/main.py", line 205, in main
    shell.execer.exec(code, mode='exec', glbs=shell.ctx)
  File "/usr/lib/python3.4/site-packages/xonsh/execer.py", line 126, in exec
    return exec(code, glbs, locs)
  File "<stdin>", line 1, in <module>
[output of ls appears here]
  File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 726, in subproc_uncaptured
    return run_subproc(cmds, captured=False)
  File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 653, in run_subproc
    wait_for_active_job()
  File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 113, in wait_for_active_job
    _give_terminal_to(pgrp)
  File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 82, in _give_terminal_to
    oldmask = signal.pthread_sigmask(signal.SIG_BLOCK, _block_when_giving)
AttributeError: 'module' object has no attribute 'pthread_sigmask'
```

Output from `xonfig`:

```
$ python
Python 3.4.3 (default, May  5 2015, 17:58:45)
[GCC 4.9.2] on cygwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from xonsh import xonfig
>>> print(xonfig.main([]))
+------------------+-------+
| xonsh            | 0.2.7 |
| Python           | 3.4.3 |
| PLY              | 3.7   |
| have readline    | True  |
| prompt toolkit   | 1.0.0 |
| pygments         | 2.1.3 |
| on posix         | True  |
| on linux         | False |
| on arch          | False |
| on windows       | False |
| on mac           | False |
| are root user    | False |
| default encoding | utf-8 |
+------------------+-------+
```

**Comment by scopatz** [comment]
what happens if you do exactly,

```python
import xonsh.tools
xonsh.tools.ON_WINDOWS = False
```

in the xonshrc file?

**Comment by scopatz** [comment]
I think that the problem is that it is picking up the  Windows subprocess model, when it should get the UNIX model since you are on Cygwin

**Comment by scopatz** [comment]
also is there a reliable way to detect we are on cygwin?

**Comment by OllieTerrance** [comment]
Using the xonshrc doesn't seem to make a difference (`xonfig` is already detecting posix though?).  I should also confirm I get a similar Ctrl-C traceback to the one in [this comment](#issuecomment-190899771), regardless of xonshrc.

As for detecting Cygwin, we have these:

```
>>> sys.platform
'cygwin'
>>> platform.system()
'CYGWIN_NT-10.0'
```

The latter varies by Windows version and 32/64 variant, but appears to always begin with `CYGWIN_NT-`.

**Comment by scopatz** [comment]
Ahh Great!

**Comment by OllieTerrance** [comment]
So it appears I can get it to work on Cygwin with the following:

```diff
diff --git a/xonsh/foreign_shells.py b/xonsh/foreign_shells.py
index 3d0acae..a8ef50a 100644
--- a/xonsh/foreign_shells.py
+++ b/xonsh/foreign_shells.py
@@ -244,6 +244,4 @@ def foreign_shell_data(shell, interactive=True, login=False, envcmd=None,
     try:
         s = subprocess.check_output(cmd, stderr=subprocess.PIPE, env=currenv,
-                                    # start new session to avoid hangs
-                                    start_new_session=True,
                                     universal_newlines=True)
     except (subprocess.CalledProcessError, FileNotFoundError):
```

**Comment by scopatz** [comment]
@OllieTerrance - All apologies about the delay.  Huh... that is super annoying.  That line was put there after much struggle to enable the Jupyter shell.  Would you mind putting in a PR that turns this off for Cygwin? I'd really appreciate it!  And again, sorry about the delay.

**Comment by OllieTerrance** [comment]
I've added another fix for the readline library modifications in readline_shell.py -- for some reason Python on Cygwin sometimes just hangs completely when interacting with `lib` (both `ctypes.c_int.in_dll` and `hasattr` were triggering it).

This still doesn't fix `AttributeError: 'module' object has no attribute 'pthread_sigmask'` -- that appears to be because the `signal` library on Cygwin literally lacks that attribute.  Can't seem to find any documentation on that either...  :disappointed:

**Comment by scopatz** [comment]
Hey @OllieTerrance - what is the full traceback of the attribute error? Maybe we can prevent that somehow?

**Comment by OllieTerrance** [comment]
There isn't one:

```
User@Host ~ $ pwd
AttributeError: 'module' object has no attribute 'pthread_sigmask'
/home/User
User@Host ~ $ cd xonsh
AttributeError: 'module' object has no attribute 'pthread_sigmask'
User@Host ~/xonsh $ ls
AttributeError: 'module' object has no attribute 'pthread_sigmask'
appveyor.yml   MANIFEST.in             setup.py
binstar.yml    README.rst              tests
CHANGELOG.rst  recipe                  travis.yml
CONTRIBUTING   release.xsh             xonsh
docs           requirements-docs.txt   xonsh-in-docker.py
license        requirements-tests.txt  xontrib
logo.txt       scripts                 xontribs.json
User@Host ~ $ python
AttributeError: 'module' object has no attribute 'pthread_sigmask'
Python 3.4.3 (default, May  5 2015, 17:58:45)
[GCC 4.9.2] on cygwin
Type "help", "copyright", "credits" or "license" for more information.
User@Host ~ $ 
```

**Comment by scopatz** [comment]
Ahh you may need to set `$XONSH_SHOW_TRACEBACK = True` to see it http://xon.sh/envvars.html#xonsh-show-traceback

**Comment by OllieTerrance** [comment]
```
User@Host ~/xonsh $ ls
Traceback (most recent call last):
  File "/home/User/xonsh/xonsh/base_shell.py", line 154, in default
    run_compiled_code(code, self.ctx, None, 'single')
appveyor.yml   MANIFEST.in             setup.py
binstar.yml    README.rst              tests
CHANGELOG.rst  recipe                  travis.yml
CONTRIBUTING   release.xsh             xonsh
docs           requirements-docs.txt   xonsh-in-docker.py
license        requirements-tests.txt  xontrib
logo.txt       scripts                 xontribs.json
  File "/home/User/xonsh/xonsh/codecache.py", line 57, in run_compiled_code
    func(code, glb, loc)
  File "<xonsh-code>", line 1, in <module>
  File "/home/User/xonsh/xonsh/built_ins.py", line 671, in subproc_captured_hiddenobject
    return run_subproc(cmds, captured='hiddenobject')
  File "/home/User/xonsh/xonsh/built_ins.py", line 573, in run_subproc
    wait_for_active_job()
  File "/home/User/xonsh/xonsh/jobs.py", line 115, in wait_for_active_job
    _give_terminal_to(pgrp)
  File "/home/User/xonsh/xonsh/jobs.py", line 84, in _give_terminal_to
    oldmask = signal.pthread_sigmask(signal.SIG_BLOCK,
AttributeError: 'module' object has no attribute 'pthread_sigmask'
```

Hmm, `_give_terminal_to` sounds like something that may explain why interactive programs aren't very interactive...

**Comment by scopatz** [comment]
I should have known that this was a job control issue.  @GuillaumeLeclerc, any ideas?

**Comment by GuillaumeLeclerc** [comment]
It's very strange. I will try to reproduce it. What are the exact versions. Maybe it's something related to cygwin

**Comment by OllieTerrance** [comment]
I suspect Python on Cygwin is at fault, it doesn't seem to provide `signal.pthread_sigmask` ([the docs say available on Unix](https://docs.python.org/3.4/library/signal.html#signal.pthread_sigmask), which usually includes Cygwin but apparently not here).

Updated versions:
```
$ python
Python 3.4.3 (default, May  5 2015, 17:58:45)
[GCC 4.9.2] on cygwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from xonsh import xonfig
>>> print(xonfig.main([]))
+------------------+-------+
| xonsh            | 0.3.0 |
| Python           | 3.4.3 |
| PLY              | 3.7   |
| have readline    | True  |
| prompt toolkit   | None  |
| pygments         | None  |
| on posix         | True  |
| on linux         | False |
| on darwin        | False |
| on windows       | False |
| is superuser     | False |
| default encoding | utf-8 |
+------------------+-------+
```

Bonus traceback, when quitting:
```
User@Host ~/xonsh $ Traceback (most recent call last):
  File "/home/User/xonsh/xonsh/proc.py", line 335, in wrapped_simple_command
    r = f(args, i)
  File "/home/User/xonsh/xonsh/aliases.py", line 146, in exit
    kill_all_jobs()
  File "/home/User/xonsh/xonsh/jobs.py", line 206, in kill_all_jobs
    _clear_dead_jobs()
  File "/home/User/xonsh/xonsh/jobs.py", line 155, in _clear_dead_jobs
    for tid in tasks:
RuntimeError: deque mutated during iteration
Traceback (most recent call last):
  File "/home/User/xonsh/xonsh/base_shell.py", line 154, in default
    run_compiled_code(code, self.ctx, None, 'single')
  File "/home/User/xonsh/xonsh/codecache.py", line 57, in run_compiled_code
    func(code, glb, loc)
  File "<xonsh-code>", line 1, in <module>
  File "/home/User/xonsh/xonsh/built_ins.py", line 671, in subproc_captured_hiddenobject
    return run_subproc(cmds, captured='hiddenobject')
  File "/home/User/xonsh/xonsh/built_ins.py", line 573, in run_subproc
    wait_for_active_job()
  File "/home/User/xonsh/xonsh/jobs.py", line 109, in wait_for_active_job
    _give_terminal_to(_shell_pgrp)  # give terminal back to the shell
  File "/home/User/xonsh/xonsh/jobs.py", line 84, in _give_terminal_to
    oldmask = signal.pthread_sigmask(signal.SIG_BLOCK,
AttributeError: 'module' object has no attribute 'pthread_sigmask'
```

**Comment by adqm** [comment]
This is almost certainly the fault of Cygwin's Python.  I am not around a Windows machine to test, but we can try removing those lines (the "block when giving" ones) and see whether things still work...

**Comment by OllieTerrance** [comment]
As it happens, that may actually be enough:

```diff
diff --git a/xonsh/jobs.py b/xonsh/jobs.py
index 9255215..82fbd15 100644
--- a/xonsh/jobs.py
+++ b/xonsh/jobs.py
@@ -82,8 +82,10 @@ else:
         # this will give the terminal to the process group pgid
         if _shell_tty is not None and os.isatty(_shell_tty):
-            oldmask = signal.pthread_sigmask(signal.SIG_BLOCK,
-                                             _block_when_giving)
             os.tcsetpgrp(_shell_tty, pgid)
-            signal.pthread_sigmask(signal.SIG_SETMASK, oldmask)

     # check for shell tty
```

This looks like it works (I can use `python` interactively!), however when run from within bash, the end of each command causes xonsh to background itself, returning to bash.  A `fg` from there puts me back at the xonsh prompt.  This doesn't seem to happen on Linux (testing with Arch here), but arguably that is down to interference from bash over anything else.

**Comment by scopatz** [comment]
I believe that `SIG_BLOCK` delays *most* signals from reaching a process. See [here](https://support.sas.com/documentation/onlinedoc/sasc/doc750/html/lr1/zlocking.htm). This is probably why it looked like it was being backgrounded.

Would you mind putting in a PR that implements a version of this function that is cygwin-specific? I'd appreciate it, since it seems like you are the only active cygwin user / tester @OllieTerrance :)

**Comment by ellocogato** [comment]
I am very pleased to have xonsh running under cygwin using both patches from @OllieTerrance (the first is already merged into xonsh; I applied the second myself) and using `exec xonsh` from bash as indicated in PR 1031.  The one issue that I see is that there is a significant delay with every command issued.  Every command seems to have about a 3 second delay after the command completes.  Subprocess commands additionally seem to have about a 3 second delay before the command is executed.

For example, issuing `(1 + 1)` immediately prints `2`, then there is a 3 second delay before the prompt appears.  When issuing `ls` in a directory with one file, there is a 3 second delay, then the filename is printed, then there is another 3 second delay before the prompt appears.

**Comment by adqm** [comment]
Great!  @ellocogato, can you try, just as a test, setting `$PROMPT = '> '`, and seeing whether that speeds things up for you?

**Comment by ellocogato** [comment]
@adqm Indeed, that did speed things up for me.  There is still a delay with anything that reads the filesystem, I think.  For example, an `ls` or `pwd` still have a delay before outputting.  Presumably it is the pwd part of the prompt that is causing a delay for all commands.  I'm curious whether @OllieTerrance is having this issue as well, or is it just me?

**Comment by adqm** [comment]
@ellocogato, I think it is actually probably those pieces of the prompt related to information about version control information, which can hit the filesystem hard.

Can you try the following?
`$PROMPT = '{env_name}{BOLD_GREEN}{user}@{hostname} {BOLD_BLUE}{cwd} {prompt_end}{NO_COLOR} '`

I imagine that wouldn't be noticeably slower than the `>` prompt.

**Comment by adqm** [comment]
FWIW, we may be able to hack in some support for the signal mask using libc's `sigprocmask` via `ctypes`.

I couldn't find much except a bug report from 2003, but since `pthread_sigmask` doesn't work (I suspect it's the Cygwin Python's fault, since Cygwin itself claims to support it), but I think it might be worth trying that as a workaround (as I think in the long term, being able to run xonsh as the default cygwin shell would be nice).

**Comment by ellocogato** [comment]
@adqm Indeed, you are correct.  That prompt has the same performance as `>`.  It definitely seems related to the filesystem.  `ls`, `cat`, filename tab completion, etc. all evoke the delay.

**Comment by ellocogato** [comment]
No clue if this is related to my performance issues, but if I sit in a directory and do `ls` over and over, about 1 in 8 times I get a ProcessLookupError:

```
Traceback (most recent call last):
  File "/usr/lib/python3.4/site-packages/xonsh/base_shell.py", line 154, in default
    run_compiled_code(code, self.ctx, None, 'single')
  File "/usr/lib/python3.4/site-packages/xonsh/codecache.py", line 57, in run_compiled_code
    func(code, glb, loc)
  File "<xonsh-code>", line 2, in <module>
  File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 607, in subproc_captured_hiddenobject
    return run_subproc(cmds, captured='hiddenobject')
  File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 509, in run_subproc
    wait_for_active_job()
  File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 130, in wait_for_active_job
    _continue(active_task)
  File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 74, in _continue
   _send_signal(job, signal.SIGCONT)
  File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 29, in _send_signal
    os.killpg(job['pgrp'], signal)
ProcessLookupError: [Errno 3] No such process
```

**Comment by OllieTerrance** [comment]
Not really seeing those delays here -- about half a second with shiny prompt, a much smaller (slightly noticeable) delay using `>`.

I am also seeing `ProcessLookupError`, but only for built-ins (i.e. `pwd`, not `ls`, nor any aliases for `ls` like `ll`).  Appears to happen every time:

```
> pwd
/home/User/xonsh
Traceback (most recent call last):
  File "/home/User/xonsh/xonsh/base_shell.py", line 154, in default
    run_compiled_code(code, self.ctx, None, 'single')
  File "/home/User/xonsh/xonsh/codecache.py", line 57, in run_compiled_code
    func(code, glb, loc)
  File "<xonsh-code>", line 1, in <module>
  File "/home/User/xonsh/xonsh/built_ins.py", line 671, in subproc_captured_hiddenobject
    return run_subproc(cmds, captured='hiddenobject')
  File "/home/User//xonsh/xonsh/built_ins.py", line 573, in run_subproc
    wait_for_active_job()
  File "/home/User/xonsh/xonsh/jobs.py", line 119, in wait_for_active_job
    _continue(active_task)
  File "/home/User/xonsh/xonsh/jobs.py", line 58, in _continue
    _send_signal(job, signal.SIGCONT)
  File "/home/User/xonsh/xonsh/jobs.py", line 64, in _send_signal
    os.killpg(job['pgrp'], signal)
ProcessLookupError: [Errno 3] No such process
>
```

**Comment by adqm** [comment]
For that error, we may need the same fix we used for Macs (see #1106) here as well, or maybe just a try/except around that call to killpg?

I think it would also be nice to have the sigmask stuff working, but neither of those blocks #1031.

**Comment by adqm** [comment]
I can't reproduce the `ProcessLookupError` issue.  Are you running the most recent version of Cygwin?

I'm going to try to figure out the sigmask stuff today, I think.

**Comment by ellocogato** [comment]
@adqm I have recently updated my Cygwin install.  DLL version is 2.5.1.

**Comment by adqm** [comment]
@ellocogato, and you're still seeing that `ProcessLookupError`?

**Comment by ellocogato** [comment]
@adqm Yes, I still see it.  I am currently setting up a clean Cygwin environment on a different machine to test xonsh, and I'll report whether that issue shows up there as well.

**Comment by adqm** [comment]
@ellocogato, @OllieTerrance: I just put in #1192, which for me fixes the issue with xonsh backgrounding itself.  It required hooking in to Cygwin's `libc` implementation via `ctypes` and it took longer than I would care to admit, because of some stupid assumptions on my part (I assumed some constants would be the same under Cygwin and GNU/Linux, but they aren't...)

but anyway, it's in!  Would you mind giving it a look?

I still haven't been able to reproduce the issue with the `ProcessLookupError`, but that is still on my list.  Have either of you had any luck with finding something that consistently reproduces it?

**Comment by jivanyatra** [comment]
I'm also attempting to use this on Cygwin (x64). After the 0.3.4 update, I run xonsh, then run pwd, and it crashes:

`Yatri@unnamed ~ $ pwd
/home/Yatri

[1]+  Stopped                 xonsh
150 Yatri@unnamed:~$
`

`fg` brings xonsh back to the foreground

When I try to get my xonfig:

`150 Yatri@unnamed:~$ python3
Python 3.4.3 (default, May  5 2015, 17:58:45)
[GCC 4.9.2] on cygwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from xonsh import xonfig
>>> print(xonfig.main([]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.4/site-packages/xonsh/xonfig.py", line 468, in main
    return _MAIN_ACTIONS[ns.action](ns)
  File "/usr/lib/python3.4/site-packages/xonsh/xonfig.py", line 339, in _info
    ('shell type', builtins.__xonsh_env__.get('SHELL_TYPE')),
AttributeError: 'module' object has no attribute '__xonsh_env__'
>>>
`

Not sure what I did wrong. It sort of worked with 0.3.3 (huge delays upwards of half a second to a second after running `pwd` or `ls`), but after the update, this happens after every command I've tried.

**Comment by adqm** [comment]
Hi @jivanyatra!  There is a workaround for the `fg` issue described [here](http://xon.sh/windows.html#running-under-cygwin) (and that's the issue that #1192 should fix).

As for the other issue...it's hard to say.  I think the `xonfig` stuff really is intended to be run from within xonsh itself, rather than from a separate Python process.  So I would say the easiest way to get your `xonfig` would be to `exec xonsh` and then run `xonfig` from there.

**Comment by jivanyatra** [comment]
Thanks @adqm !! So, `exec xonsh` now works, and `xonfig` outputs:
    +------------------+----------------+
    | xonsh            | 0.3.4          |
    | Python           | 3.4.3          |
    | PLY              | 3.7            |
    | have readline    | True           |
    | prompt toolkit   | 1.0.0          |
    | shell type       | prompt_toolkit |
    | pygments         | 2.1.3          |
    | on posix         | True           |
    | on linux         | False          |
    | on darwin        | False          |
    | on windows       | False          |
    | on cygwin        | True           |
    | is superuser     | False          |
    | default encoding | utf-8          |
    +------------------+----------------+

I saw that I didn't have prompt_toolkit or pygments installed, so I installed them. Now, I get this result from `pwd`:
    Yatri@unnamed ~ $ pwd
          0 [main] python3 12124 fork: child -1 - CreateProcessW failed for '', errno 2
    xonsh: subprocess mode: command not found: pwd
    Did you mean one of the following?
        cd:    Alias
        pip:   Command (/usr/bin/pip)
        popd:  Alias
        awk:   Command (/usr/bin/awk)
        mad:   Command (/usr/bin/mad)
    2219966 [main] python3 12124 fork: child -1 - CreateProcessW failed for '', errno 2

**Comment by adqm** [comment]
@jivanyatra, thanks!  That `ProcessLookupError` is the one I haven't yet been able to reproduce on my own machines.  But thanks for this information.  I can maybe try running with prompt_toolkit to see if that is somehow responsible (I had been using the `readline` shell), but other than that, our configurations look pretty much the same.  Anyway, I'll keep trying that.   And #1192 should make it into the next release, so we'll at least be making little steps toward full Cygwin support!

**Comment by jivanyatra** [comment]
If there's anything else I can provide, I'm happy to help. I'm a new python dev and this project definitely caught my attention. I'd love to contribute as well, but I'm not sure what I could do with my admittedly novice skill level.

**Comment by ellocogato** [comment]
@adqm I tested #1192 and it looks like it properly fixes the issue of xonsh backgrounding itself.

**Comment by ellocogato** [comment]
@adqm Here's an interesting twist.  I still get the occasional `ProcessLookupError`, and now (with #1192 applied), xonsh will background itself whenever the error occurs.

**Comment by adqm** [comment]
@ellocogato: Weird!  Okay, so maybe this makes it even more crucial to fix that issue.  I wish I could reproduce it personally (I still haven't been able to).

Is that backgrounding also fixed with `exec xonsh`, or is it actually _more_ broken than before?

**Comment by ellocogato** [comment]
@adqm OK, one more thing to throw in the mix.  Not sure if this was happening before #1192, but I just noticed that sometimes instead of a `ProcessLookupError`, I get a `PermissionError`:

```
Traceback (most recent call last):
  File "/usr/lib/python3.4/site-packages/xonsh/base_shell.py", line 154, in default
    run_compiled_code(code, self.ctx, None, 'single')
  File "/usr/lib/python3.4/site-packages/xonsh/codecache.py", line 57, in run_compiled_code
    func(code, glb, loc)
  File "<xonsh-code>", line 1, in <module>
  File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 607, in subproc_captured_hiddenobject
    return run_subproc(cmds, captured='hiddenobject')
  File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 509, in run_subproc
    wait_for_active_job()
  File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 153, in wait_for_active_job
    _continue(active_task)
  File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 74, in _continue
    _send_signal(job, signal.SIGCONT)
  File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 29, in _send_signal
    os.killpg(job['pgrp'], signal)
PermissionError: [Errno 13] Permission denied
```

**Comment by adqm** [comment]
I feel like that had been reported elsewhere as well, and I think I know of a likely fix.  I'll try to put in another PR tonight.  Do you mind helping to test if I'm not able to reproduce myself?

**Comment by ellocogato** [comment]
Regarding `exec xonsh`, after #1192 everything appears to behave the same as before.  No backgrounding during normal operation or when an error occurs.

**Comment by adqm** [comment]
Alright, great!  Then we've just got that one last thing to track down, it seems, before this works as expected on Cygwin.  My hopes are high!

**Comment by jivanyatra** [comment]
Okay, after more testing, I'm seeing the same thing as @ellocogato. And there is no more backgrounding occurring.

**Comment by adqm** [comment]
Great; thanks for all the help!

**Comment by vmiheer** [comment]
This was long thread so haven't looked at all of it.
I can run xonsh on cygwin64 python 3.4.
Problem I see is when I type `ls` command it shows the output and then xonsh goes to background bringing me back to bash.
Doing `fg` in bash brings me back to xonsh but only till I hit `ls` again (which is so prevalent).

**Comment by OllieTerrance** [comment]
This seems to be working much better now.  :smile:  Only thing I'm seeing now is the 2-3 second delay redrawing the prompt, which was mentioned earlier -- is anyone else still seeing this?

(`$PROMPT = '> '` removes the delay, so I guess it's the prompt formatting here.  It's also a bit slow on my Arch RPi, but I'm assuming the Pi is to blame for that.)

**Comment by adqm** [comment]
@vmiheer, yes this is a known issue that should be fixed in the next release (via #1192 and #1193).  As a workaround in the meantime, you should be able to use `exec xonsh` instead of just `xonsh` to open the shell.

@OllieTerrance: Great!  Glad to hear it is working now.  We will get this into the next release for sure.  It is also true that #1166 (also going in to the next release) will remove the VCS branch information from the prompt, which should speed things up drastically while still giving a somewhat nice looking prompt.  As for the RPi, xonsh is pretty heavy-weight for a shell, so I expect there isn't much we can do about it on the pi.

**Comment by ellocogato** [comment]
@adqm PR #1193 does prevent the occasional `ProcessLookupError` and `PermissionError`.

But now I get some new ones!  I have seen a `BlockingIOError` (just once so far), `OSError`, and `BrokenPipeError`.  These all occur in `os.killpg()`.  If I handle `BlockingIOError` and `OSError` in addition to the others, then that seems to cover it (I guess BrokenPipeError is a subclass of one of the other two).

Also note that I put some debugging output in this error handling code, and I noticed that every time `killpg` fails, the `kill` also fails.  (`killpg` can be any of the four errors, but `kill` is always a `ProcessLookupError`.)  So maybe the `kill` part is unnecessary?  If you want to see this in action: http://pastebin.com/A7NWe6QT


```
Traceback (most recent call last):
  File "/usr/lib/python3.4/site-packages/xonsh/base_shell.py", line 154, in default
    run_compiled_code(code, self.ctx, None, 'single')
  File "/usr/lib/python3.4/site-packages/xonsh/codecache.py", line 57, in run_compiled_code
    func(code, glb, loc)
  File "<xonsh-code>", line 1, in <module>
  File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 607, in subproc_captured_hiddenobject
    return run_subproc(cmds, captured='hiddenobject')
  File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 509, in run_subproc
    wait_for_active_job()
  File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 170, in wait_for_active_job
    _continue(active_task)
  File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 91, in _continue
    _send_signal(job, signal.SIGCONT)
  File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 33, in _send_signal
    os.killpg(job['pgrp'], signal)
BlockingIOError: [Errno 11] Resource temporarily unavailable
 
[1]+  Stopped                 xonsh
```

```
Traceback (most recent call last):
  File "/usr/lib/python3.4/site-packages/xonsh/base_shell.py", line 154, in default
    run_compiled_code(code, self.ctx, None, 'single')
  File "/usr/lib/python3.4/site-packages/xonsh/codecache.py", line 57, in run_compiled_code
    func(code, glb, loc)
  File "<xonsh-code>", line 1, in <module>
  File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 585, in subproc_captured_stdout
    return run_subproc(cmds, captured='stdout')
  File "/usr/lib/python3.4/site-packages/xonsh/built_ins.py", line 509, in run_subproc
    wait_for_active_job()
  File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 168, in wait_for_active_job
    _continue(active_task)
  File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 89, in _continue
    _send_signal(job, signal.SIGCONT)
  File "/usr/lib/python3.4/site-packages/xonsh/jobs.py", line 32, in _send_signal
    os.killpg(job['pgrp'], signal)
OSError: [Errno 22] Invalid argument
 
[1]+  Stopped                 xonsh
```

**Comment by adqm** [comment]
@ellocogato: just updated the patch.  do you mind giving it another shot?  thanks for all the help with testing this!

**Comment by ellocogato** [comment]
@adqm: the update to the patch seems to completely resolve the issue.  Is there any consequence of the fact that these errors are occurring (and that they are now being ignoring)?

**Comment by adqm** [comment]
@ellocogato: maybe... i guess in the worst case, there could still be a process floating around someplace.  i'm not too worried about that, though, since i would imagine the only way we could hit some of these errors is if the processes are already dead and gone...

**Comment by adqm** [comment]
Alright, I think we've mostly taken care of these issues.  I'm going to close for now.  Feel free to re-open (or to open another issue) if problems persist under Cygwin.
